pkg/cmd/server/etcd/etcdserver/run.go

```diff
@@ -1,64 +1,95 @@
-		glog.Fatalf("Unable to build etcd client URLs: %v", err)
+	u, err := types.NewURLs(addressToURLs(etcdServerConfig.ServingInfo.BindAddress, clientTLS))
+	if err != nil {
+		glog.Fatalf("Unable to build etcd peer URLs: %v", err)
 	}
+	cfg.LCUrls = []url.URL(u)
 
-	if configapi.UseTLS(etcdServerConfig.PeerServingInfo) {
-		cfg.peerTLSInfo.CAFile = etcdServerConfig.PeerServingInfo.ClientCA
-		cfg.peerTLSInfo.CertFile = etcdServerConfig.PeerServingInfo.ServerCert.CertFile
-		cfg.peerTLSInfo.KeyFile = etcdServerConfig.PeerServingInfo.ServerCert.KeyFile
+	peerTLS := configapi.UseTLS(etcdServerConfig.PeerServingInfo)
+	if peerTLS {
+		cfg.PeerTLSInfo.CAFile = etcdServerConfig.PeerServingInfo.ClientCA
+		cfg.PeerTLSInfo.CertFile = etcdServerConfig.PeerServingInfo.ServerCert.CertFile
+		cfg.PeerTLSInfo.KeyFile = etcdServerConfig.PeerServingInfo.ServerCert.KeyFile
+		cfg.PeerTLSInfo.ClientCertAuth = len(cfg.PeerTLSInfo.CertFile) > 0
 	}
-	if cfg.lpurls, err = urlsFromStrings(etcdServerConfig.PeerServingInfo.BindAddress, cfg.peerTLSInfo); err != nil {
+	u, err = types.NewURLs(addressToURLs(etcdServerConfig.PeerServingInfo.BindAddress, peerTLS))
+	if err != nil {
 		glog.Fatalf("Unable to build etcd peer URLs: %v", err)
 	}
+	cfg.LPUrls = []url.URL(u)
 
-	if cfg.acurls, err = urlsFromStrings(etcdServerConfig.Address, cfg.clientTLSInfo); err != nil {
+	u, err = types.NewURLs(addressToURLs(etcdServerConfig.Address, clientTLS))
+	if err != nil {
 		glog.Fatalf("Unable to build etcd announce client URLs: %v", err)
 	}
-	if cfg.apurls, err = urlsFromStrings(etcdServerConfig.PeerAddress, cfg.peerTLSInfo); err != nil {
-		glog.Fatalf("Unable to build etcd announce peer URLs: %v", err)
-	}
+	cfg.ACUrls = []url.URL(u)
 
-	if err := cfg.resolveUrls(); err != nil {
-		glog.Fatalf("Unable to resolve etcd URLs: %v", err)
+	u, err = types.NewURLs(addressToURLs(etcdServerConfig.PeerAddress, peerTLS))
+	if err != nil {
+		glog.Fatalf("Unable to build etcd announce peer URLs: %v", err)
 	}
+	cfg.APUrls = []url.URL(u)
 
-	cfg.initialCluster = fmt.Sprintf("%s=%s", cfg.name, cfg.apurls[0].String())
+	cfg.InitialCluster = cfg.InitialClusterFromName(cfg.Name)
 
-	stopped, err := startEtcd(cfg)
+	e, err := embed.StartEtcd(cfg)
 	if err != nil {
 		glog.Fatalf("Unable to start etcd: %v", err)
 	}
+
 	go func() {
-		glog.Infof("Started etcd at %s", etcdServerConfig.Address)
-		<-stopped
+		defer e.Close()
+
+		select {
+		case <-e.Server.ReadyNotify():
+			glog.Infof("Started etcd at %s", etcdServerConfig.Address)
+		case <-time.After(60 * time.Second):
+			glog.Warning("etcd took too long to start, stopped")
+			e.Server.Stop() // trigger a shutdown
+		}
+		glog.Fatalf("etcd has returned an error: %v", <-e.Err())
 	}()
 }
+
+// addressToURLs turns a host:port comma delimited list into an array valid
+// URL strings with the appropriate prefix for the TLS mode.
+func addressToURLs(addr string, isTLS bool) []string {
```

Comment:
we're passing in bind addresses here... is `https://0.0.0.0:4001` going to be valid?